state/protocol/badger/mutator.go

```diff
@@ -1121,6 +1121,11 @@ func (m *FollowerState) handleEpochServiceEvents(candidate *flow.Block, updater
 
 				err = updater.ProcessEpochSetup(ev)
 				if err != nil {
+					if protocol.IsInvalidServiceEventError(err) {
+						// we have observed an invalid service event, which triggers epoch fallback mode
+						updater.SetInvalidStateTransitionAttempted()
+						return dbUpdates, nil
+					}
```

Comment:
## ⚠️ 

I am a little worried about this: 
At this point, we don't know that the `updater`'s internal state is a valid Protocol State. For all we know, it could be complete garbage. There is no atomicity requirement on the `StateUpdater` as far as I can tell, in that it either applies the update entirely or not at all. 

   Suggestion:
   * I think it it is ok to treat all protocol state operations mutations in a block as one atomic state transition. It either succeeds in its entirety or in case of a failure, there is no update to the protocol state (except setting `InvalidStateTransitionAttempted` to `true`).
   * I think it is straight forward to extend the protocol state `Updater` to "drop all modifications if `InvalidStateTransitionAttempted`" and document this properly throughout the code base. In a nutshell, within the [`Build` method](https://github.com/onflow/flow-go/blob/0172dd130c724b60f94787dd2c1f54b9cd7cc0e0/state/protocol/protocol_state/updater.go#L47-L53) I would check the `state.InvalidStateTransitionAttempted` first; in that case we could just return a copy of the [`parentState`](https://github.com/onflow/flow-go/blob/0172dd130c724b60f94787dd2c1f54b9cd7cc0e0/state/protocol/protocol_state/updater.go#L21) with only the `state.InvalidStateTransitionAttempted` set to true. 
   * While the code changes are relatively small, we need to properly test this and add detailed documentation to the `Updater`, the corresponding interface, and ideally also the places where we call the `Updater`

   Further thoughts: 
   * If the updater is signalling that it encountered an invalid state transition via an `InvalidStateTransitionAttempted`, should the updater maybe already set its own `InvalidStateTransitionAttempted` flag? We would still raise the error to signal the failure to the caller. However, the caller would no longer be required to explicitly call `updater.SetInvalidStateTransitionAttempted()` (I'd leave that method, to allow `Updater` external logic to set this flag for other reasons).

I don't think we should do this as part of the current PR. The PR is already big enough and we keep layering on changes which makes it hard to keep track of without re-reviewing the entire PR.